pkg/cmd/cli/cmd/version.go

```diff
@@ -0,0 +1,163 @@
+package cmd
+
+import (
+	"encoding/json"
+	"fmt"
+	"io"
+	"strings"
+	"time"
+
+	etcdversion "github.com/coreos/etcd/version"
+
+	kcmdutil "k8s.io/kubernetes/pkg/kubectl/cmd/util"
+	kubeversion "k8s.io/kubernetes/pkg/version"
+
+	"github.com/openshift/origin/pkg/cmd/util/clientcmd"
+	"github.com/openshift/origin/pkg/cmd/util/tokencmd"
+	"github.com/openshift/origin/pkg/version"
+
+	"github.com/spf13/cobra"
+)
+
+const (
+	versionLong = `
+Display client and server versions.`
+)
+
+type VersionOptions struct {
+	BaseName string
+	Out      io.Writer
+
+	PrintEtcdVersion    bool
+	PrintClientFeatures bool
+}
+
+// NewCmdVersion creates a command for displaying the version of this binary
+func NewCmdVersion(fullName string, f *clientcmd.Factory, out io.Writer, options VersionOptions) *cobra.Command {
+	cmd := &cobra.Command{
+		Use:   "version",
+		Short: "Display client and server versions",
+		Long:  versionLong,
+		Run: func(cmd *cobra.Command, args []string) {
+			options.BaseName = fullName
+
+			if err := options.Complete(args, out); err != nil {
+				kcmdutil.CheckErr(kcmdutil.UsageError(cmd, err.Error()))
+			}
+
+			if err := options.RunVersion(f); err != nil {
+				kcmdutil.CheckErr(err)
+			}
+		},
+	}
+
+	return cmd
+}
+
+func (o *VersionOptions) Complete(args []string, out io.Writer) error {
+	if len(args) > 0 {
+		return fmt.Errorf("no arguments should be passed")
+	}
+	o.Out = out
+
+	return nil
+}
+
+// RunVersion attempts to display client and server versions for Kubernetes and OpenShift
+func (o VersionOptions) RunVersion(f *clientcmd.Factory) error {
```

Comment:
he means the signature should be `RunVersion()`, no factory passed in. You can pass the factory into Complete() and grab what you need from there. If you don't want to actually invoke the factory method to get the clients until you get to `RunVersion()`, you can add a `Clients func()` to the Options struct, and set it to the factory method in `Complete()`.